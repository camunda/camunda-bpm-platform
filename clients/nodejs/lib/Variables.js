/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  getVariableType,
  mapEntries,
  serializeVariable,
  deserializeVariable,
} from "./__internal/utils.js";

function Variables(initialVariables = {}, options = {}) {
  const { readOnly, processInstanceId, engineService } = options;

  let dirtyVariables = {};

  /**
   * @returns the typedValue corresponding to variableName
   * @param variableName
   */
  this.getTyped = (variableName) => {
    let typedValue = initialVariables[variableName];

    if (!typedValue) {
      return null;
    }

    return deserializeVariable({
      key: variableName,
      typedValue,
      processInstanceId,
      engineService,
    });
  };

  /**
   * @returns the value corresponding to variableName
   * @param variableName
   */
  this.get = (variableName) => {
    const { value } = { ...this.getTyped(variableName) };
    return value;
  };

  /**
   * @returns the values of all variables
   */
  this.getAll = () =>
    mapEntries(initialVariables, ({ key }) => ({ [key]: this.get(key) }));

  /**
   * @returns the typed values of all variables
   */
  this.getAllTyped = () =>
    mapEntries(initialVariables, ({ key }) => ({ [key]: this.getTyped(key) }));

  /**
   * @returns the dirty variables
   */
  this.getDirtyVariables = () => {
    return dirtyVariables;
  };

  if (!readOnly) {
    /**
     * Sets typed value for variable corresponding to variableName
     * @param variableName
     * @param typedValue
     */
    this.setTyped = (variableName, typedValue) => {
      initialVariables[variableName] = dirtyVariables[variableName] =
        serializeVariable({
          key: variableName,
          typedValue,
        });
      return this;
    };

    /**
     * Sets value for variable corresponding to variableName
     * The type is determined automatically
     * @param variableName
     * @param value
     */
    this.set = (variableName, value) => {
      const type = getVariableType(value);
      return this.setTyped(variableName, { type, value, valueInfo: {} });
    };

    /**
     * Sets value for variable corresponding to variableName
     * The type is determined automatically
     * The variable has transient flag: true
     * @param variableName
     * @param value
     */
    this.setTransient = (variableName, value) => {
      const type = getVariableType(value);
      return this.setTyped(variableName, {
        type,
        value,
        valueInfo: { transient: true },
      });
    };

    /**
     * Sets the values of multiple variables at once
     * The new values are merged with existing ones
     * @param values
     */
    this.setAll = (values) => {
      const self = this;
      Object.entries(values).forEach(([key, value]) => {
        self.set(key, value);
      });
      return self;
    };

    /**
     * Sets the typed values of multiple variables at once
     * The new typedValues are merged with existing ones
     * @param typedValues
     */
    this.setAllTyped = (typedValues) => {
      const self = this;
      Object.entries(typedValues).forEach(([key, typedValue]) => {
        self.setTyped(key, typedValue);
      });
      return self;
    };
  }
}

export default Variables;
