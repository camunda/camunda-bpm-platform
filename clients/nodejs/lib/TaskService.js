/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  MISSING_TASK,
  MISSING_ERROR_CODE,
  MISSING_DURATION,
  MISSING_NEW_DURATION,
} from "./__internal/errors.js";
import { isUndefinedOrNull } from "./__internal/utils.js";
import Variables from "./Variables.js";

class TaskService {
  constructor(events, api) {
    this.events = events;
    this.api = api;

    this.sanitizeTask = this.sanitizeTask.bind(this);
    this.success = this.success.bind(this);
    this.error = this.error.bind(this);
    this.complete = this.complete.bind(this);
    this.handleFailure = this.handleFailure.bind(this);
    this.handleBpmnError = this.handleBpmnError.bind(this);
    this.lock = this.lock.bind(this);
    this.extendLock = this.extendLock.bind(this);
  }

  sanitizeTask(task) {
    if (typeof task === "object") {
      return { id: task.id };
    } else {
      return { id: task };
    }
  }

  success(event, ...args) {
    this.events.emit(`${event}:success`, ...args);
  }

  error(event, ...args) {
    this.events.emit(`${event}:error`, ...args);
  }

  /**
   * @throws Error
   * @param task
   * @returns {Promise<void>}
   */
  async complete(task, variables, localVariables) {
    if (isUndefinedOrNull(task)) {
      throw new Error(MISSING_TASK);
    }

    const sanitizedTask = this.sanitizeTask(task);
    try {
      const requestBody = { ...sanitizedTask };
      if (variables instanceof Variables) {
        requestBody.variables = variables.getDirtyVariables();
      }
      if (localVariables instanceof Variables) {
        requestBody.localVariables = localVariables.getDirtyVariables();
      }
      await this.api.complete(requestBody);
      this.success("complete", task);
    } catch (e) {
      this.error("complete", task, e);
      throw e;
    }
  }

  /**
   * @throws Error
   * @param task
   * @param options
   * @returns {Promise<void>}
   */
  async handleFailure(task, options) {
    if (isUndefinedOrNull(task)) {
      throw new Error(MISSING_TASK);
    }

    const sanitizedTask = this.sanitizeTask(task);
    try {
      await this.api.handleFailure(sanitizedTask, options);
      this.success("handleFailure", task);
    } catch (e) {
      this.error("handleFailure", task, e);
      throw e;
    }
  }

  /**
   * @throws Error
   * @param task
   * @param errorCode
   * @param errorMessage
   * @param variables
   * @returns {Promise<void>}
   */
  async handleBpmnError(task, errorCode, errorMessage, variables) {
    if (isUndefinedOrNull(task)) {
      throw new Error(MISSING_TASK);
    }
    if (!errorCode) {
      throw new Error(MISSING_ERROR_CODE);
    }

    const sanitizedTask = this.sanitizeTask(task);

    try {
      await this.api.handleBpmnError(
        sanitizedTask,
        errorCode,
        errorMessage,
        variables instanceof Variables
          ? variables.getDirtyVariables()
          : undefined
      );
      this.success("handleBpmnError", task);
    } catch (e) {
      this.error("handleBpmnError", task, e);
      throw e;
    }
  }

  /**
   * @throws Error
   * @param task
   * @param duration
   * @returns {Promise<void>}
   */
  async lock(task, duration) {
    if (isUndefinedOrNull(task)) {
      throw new Error(MISSING_TASK);
    }
    if (!duration) {
      throw new Error(MISSING_DURATION);
    }

    const sanitizedTask = this.sanitizeTask(task);
    try {
      await this.api.lock(sanitizedTask, duration);
      this.success("lock", task);
    } catch (e) {
      this.error("lock", task, e);
      throw e;
    }
  }

  /**
   * @throws Error
   * @param task
   * @param newDuration
   * @returns {Promise<void>}
   */
  async extendLock(task, newDuration) {
    if (isUndefinedOrNull(task)) {
      throw new Error(MISSING_TASK);
    }
    if (!newDuration) {
      throw new Error(MISSING_NEW_DURATION);
    }

    const sanitizedTask = this.sanitizeTask(task);
    try {
      await this.api.extendLock(sanitizedTask, newDuration);
      this.success("extendLock", task);
    } catch (e) {
      this.error("extendLock", task, e);
      throw e;
    }
  }

  /**
   * @throws Error
   * @param task
   * @returns {Promise<void>}
   */
  async unlock(task) {
    if (isUndefinedOrNull(task)) {
      throw new Error(MISSING_TASK);
    }

    const sanitizedTask = this.sanitizeTask(task);
    try {
      await this.api.unlock(sanitizedTask);
      this.success("unlock", task);
    } catch (e) {
      this.error("unlock", task, e);
      throw e;
    }
  }
}

export default TaskService;
