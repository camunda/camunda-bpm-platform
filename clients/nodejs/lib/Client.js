/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import events from "events";

import EngineService from "./__internal/EngineService.js";
import TaskService from "./TaskService.js";
import Variables from "./Variables.js";

import {
  MISSING_BASE_URL,
  ALREADY_REGISTERED,
  MISSING_HANDLER,
  WRONG_INTERCEPTOR,
  WRONG_MIDDLEWARES,
  WRONG_SORTING,
  WRONG_SORTING_SORT_ORDER,
  WRONG_SORTING_SORT_BY,
} from "./__internal/errors.js";

import {
  isFunction,
  isArrayOfFunctions,
  isUndefinedOrNull,
} from "./__internal/utils.js";

const defaultOptions = {
  workerId: "some-random-id",
  maxTasks: 10,
  usePriority: true,
  interval: 300,
  lockDuration: 50000,
  autoPoll: true,
};

/**
 * @throws Error
 * @param customOptions
 * @param customOptions.baseUrl: baseUrl to api | REQUIRED
 * @param customOptions.workerId
 * @param customOptions.maxTasks
 * @param customOptions.interval
 * @param customOptions.lockDuration
 * @param customOptions.autoPoll
 * @param customOptions.asyncResponseTimeout
 * @param customOptions.interceptors
 * @param customOptions.usePriority
 * @param customOptions.maxParallelExecutions
 * @param customOptions.sorting
 * @constructor
 */
class Client extends events {
  static SortBy = Object.freeze({
    CreateTime: "createTime",
  });

  static SortOrder = Object.freeze({
    ASC: "asc",
    DESC: "desc",
  });

  constructor(customOptions) {
    super();

    /**
     * Bind member methods
     */
    this.sanitizeOptions = this.sanitizeOptions.bind(this);
    this.start = this.start.bind(this);
    this.poll = this.poll.bind(this);
    this.subscribe = this.subscribe.bind(this);
    this.executeTask = this.executeTask.bind(this);
    this.stop = this.stop.bind(this);

    /**
     *  Initialize data
     */
    this.activeTasksCount = 0;
    this.sanitizeOptions(customOptions);
    this.topicSubscriptions = {}; // Map which contains all subscribed topics
    this.engineService = new EngineService(this.options);
    this.taskService = new TaskService(this, this.engineService);

    if (this.options.use) {
      this.options.use.forEach((f) => f(this));
    }

    if (this.options.autoPoll) {
      this.start();
    }
  }

  /**
   * @throws Error
   * @param customOptions
   */
  sanitizeOptions(customOptions) {
    this.options = { ...defaultOptions, ...customOptions };

    if (
      !customOptions ||
      !customOptions.baseUrl ||
      !customOptions.baseUrl.length
    ) {
      throw new Error(MISSING_BASE_URL);
    }

    const { interceptors, use } = customOptions;

    // sanitize request interceptors
    if (
      interceptors &&
      !isFunction(interceptors) &&
      !isArrayOfFunctions(interceptors)
    ) {
      throw new Error(WRONG_INTERCEPTOR);
    }

    if (isFunction(interceptors)) {
      this.options.interceptors = [interceptors];
    }

    // sanitize middlewares
    if (use && !isFunction(use) && !isArrayOfFunctions(use)) {
      throw new Error(WRONG_MIDDLEWARES);
    }

    if (isFunction(use)) {
      this.options.use = [use];
    }

    if (this.options.sorting) {
      // sanitize sorting
      if (!Array.isArray(this.options.sorting)) {
        throw new Error(WRONG_SORTING);
      }

      // sanitize sorting.sortBy
      const sortByPredicate = (sorting) =>
        !sorting.sortBy ||
        !Object.values(Client.SortBy).includes(sorting.sortBy);
      if (this.options.sorting.some(sortByPredicate)) {
        throw new Error(WRONG_SORTING_SORT_BY + Object.values(Client.SortBy));
      }

      // sanitize sorting.sortOrder
      const sortOrderPredicate = (sorting) =>
        !sorting.sortOrder ||
        !Object.values(Client.SortOrder).includes(sorting.sortOrder);
      if (this.options.sorting.some(sortOrderPredicate)) {
        throw new Error(
          WRONG_SORTING_SORT_ORDER + Object.values(Client.SortOrder)
        );
      }
    }
  }

  /**
   * Starts polling
   */
  start() {
    this._isPollAllowed = true;
    this.poll();
  }

  /**
   * Polls tasks from engine and executes them
   */
  async poll() {
    if (!this._isPollAllowed) {
      return;
    }

    this.emit("poll:start");

    const { topicSubscriptions, engineService, executeTask, poll } = this;
    const {
      maxTasks,
      usePriority,
      sorting,
      interval,
      asyncResponseTimeout,
      maxParallelExecutions,
    } = this.options;

    const requiredTasksCount = isUndefinedOrNull(maxParallelExecutions)
      ? maxTasks
      : Math.min(maxTasks, maxParallelExecutions - this.activeTasksCount);
    // If all previously received tasks are running now, reschedule polling
    if (requiredTasksCount <= 0) {
      return setTimeout(poll, interval);
    }

    // setup polling options
    let pollingOptions = {
      maxTasks: requiredTasksCount,
      usePriority: usePriority,
    };
    if (asyncResponseTimeout) {
      pollingOptions = { ...pollingOptions, asyncResponseTimeout };
    }
    if (sorting) {
      pollingOptions = { ...pollingOptions, sorting };
    }

    // if there are no topic subscriptions, reschedule polling
    if (!Object.keys(topicSubscriptions).length) {
      return setTimeout(poll, interval);
    }

    // collect topics that have subscriptions
    const topics = Object.entries(topicSubscriptions).map(
      ([
        topicName,
        {
          businessKey,
          lockDuration,
          variables,
          processDefinitionId,
          processDefinitionIdIn,
          processDefinitionKey,
          processDefinitionKeyIn,
          processDefinitionVersionTag,
          processVariables,
          tenantIdIn,
          withoutTenantId,
          localVariables,
          includeExtensionProperties,
        },
      ]) => {
        let topic = { topicName, lockDuration };

        if (!isUndefinedOrNull(variables)) {
          topic.variables = variables;
        }

        if (!isUndefinedOrNull(businessKey)) {
          topic.businessKey = businessKey;
        }

        if (!isUndefinedOrNull(processDefinitionId)) {
          topic.processDefinitionId = processDefinitionId;
        }

        if (!isUndefinedOrNull(processDefinitionIdIn)) {
          topic.processDefinitionIdIn = processDefinitionIdIn;
        }

        if (!isUndefinedOrNull(processDefinitionKey)) {
          topic.processDefinitionKey = processDefinitionKey;
        }

        if (!isUndefinedOrNull(processDefinitionKeyIn)) {
          topic.processDefinitionKeyIn = processDefinitionKeyIn;
        }

        if (!isUndefinedOrNull(processDefinitionVersionTag)) {
          topic.processDefinitionVersionTag = processDefinitionVersionTag;
        }

        if (!isUndefinedOrNull(processVariables)) {
          topic.processVariables = processVariables;
        }

        if (!isUndefinedOrNull(tenantIdIn)) {
          topic.tenantIdIn = tenantIdIn;
        }

        if (!isUndefinedOrNull(withoutTenantId)) {
          topic.withoutTenantId = withoutTenantId;
        }

        if (!isUndefinedOrNull(localVariables)) {
          topic.localVariables = localVariables;
        }

        if (!isUndefinedOrNull(includeExtensionProperties)) {
          topic.includeExtensionProperties = includeExtensionProperties;
        }

        return topic;
      }
    );

    const requestBody = { ...pollingOptions, topics };

    try {
      const tasks = await engineService.fetchAndLock(requestBody);
      this.emit("poll:success", tasks);
      tasks.forEach(executeTask);
    } catch (e) {
      this.emit("poll:error", e);
    }
    setTimeout(poll, interval);
  }

  /**
   * Subscribes a handler by adding a topic subscription
   * @param topic
   * @param customOptions
   * @param customOptions.lockDuration
   * @param handler
   */
  subscribe(topic, customOptions, handler) {
    const topicSubscriptions = this.topicSubscriptions;
    const options = this.options;
    const lockDuration = options.lockDuration;

    if (topicSubscriptions[topic]) {
      throw new Error(ALREADY_REGISTERED);
    }

    // handles the case if there is no options being
    if (typeof customOptions === "function") {
      handler = customOptions;
      customOptions = null;
    }

    if (!handler) {
      throw new Error(MISSING_HANDLER);
    }
    const unsubscribe = () => {
      delete topicSubscriptions[topic];
      this.emit("unsubscribe");
    };

    const topicSubscription = {
      handler,
      unsubscribe,
      lockDuration,
      ...customOptions,
    };
    topicSubscriptions[topic] = topicSubscription;
    this.emit("subscribe", topic, topicSubscription);

    return topicSubscription;
  }

  /**
   * Executes task using the worker registered to its topic
   */
  async executeTask(task) {
    const taskService = this.taskService;
    const topicSubscription = this.topicSubscriptions[task.topicName];
    const variables = new Variables(task.variables, {
      readOnly: true,
      processInstanceId: task.processInstanceId,
      engineService: this.engineService,
    });

    this.activeTasksCount++;
    const newTask = { ...task, variables };
    try {
      await topicSubscription.handler({ task: newTask, taskService });
      this.emit("handler:success", task);
    } catch (e) {
      this.emit("handler:error", e);
    } finally {
      this.activeTasksCount--;
    }
  }

  /**
   * Stops polling
   */
  stop() {
    this._isPollAllowed = false;
    this.emit("poll:stop");
  }
}

export default Client;
