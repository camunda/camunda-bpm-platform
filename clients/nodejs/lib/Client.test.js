/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { jest } from "@jest/globals";
import Client from "./Client.js";
import TaskService from "./TaskService.js";
import Variables from "./Variables.js";

import {
  WRONG_INTERCEPTOR,
  MISSING_BASE_URL,
  ALREADY_REGISTERED,
  MISSING_HANDLER,
  WRONG_MIDDLEWARES,
} from "./__internal/errors.js";

// workaround for https://github.com/facebook/jest/issues/2157
const flushPromises = () =>
  new Promise(jest.requireActual("timers").setImmediate);
const advanceTimersByTime = async (msToRun) => {
  await flushPromises();
  return jest.advanceTimersByTime(msToRun);
};

jest.mock("got");
jest.mock("../lib/Variables");

const customClientOptions = {
  baseUrl: "http://localhost:XXXX/engine-rest/",
  workerId: "foobarId",
  maxTasks: 3,
  interval: 100,
  lockDuration: 30000,
  sorting: [
    { sortBy: Client.SortBy.CreateTime, sortOrder: Client.SortOrder.ASC },
  ],
};

describe("Client", () => {
  describe("poll ", () => {
    jest.useFakeTimers();
    describe("when autoPoll is false", () => {
      let pollSpy, client, engineService;

      beforeEach(() => {
        client = new Client({ ...customClientOptions, autoPoll: false });
        engineService = client.engineService;
        pollSpy = jest.spyOn(client, "poll");
      });

      test("calling start should call poll", () => {
        // when
        client.start();

        // then
        expect(pollSpy).toHaveBeenCalled();
      });

      test("should call itself again after timeout when there are no topic subscriptions", async () => {
        // when
        client.start();
        await advanceTimersByTime(customClientOptions.interval);

        // then
        expect(pollSpy).toHaveBeenCalledTimes(2);
      });

      test("should not call itself again after timeout when there are registered client", async () => {
        // given
        client.subscribe("foo", () => {});

        // when
        client.start();
        await advanceTimersByTime(customClientOptions.interval);

        // then
        expect(pollSpy).toHaveBeenCalledTimes(2);
      });

      test("should fetchAndLock and then call itself again when there are registered client", async () => {
        // given
        const fetchAndLockSpy = jest.spyOn(engineService, "fetchAndLock");
        client.subscribe("foo", () => {});

        // when
        client.start();
        await advanceTimersByTime(customClientOptions.interval);

        // then
        expect(fetchAndLockSpy).toBeCalled();
        expect(pollSpy).toHaveBeenCalledTimes(2);
      });
    });

    describe("when autoPoll is true", () => {
      it("should call itself automatically when autoPoll is true", async () => {
        // given
        const client = new Client({ ...customClientOptions, autoPoll: true });
        const pollSpy = jest.spyOn(client, "poll");

        // when
        await advanceTimersByTime(2 * customClientOptions.interval);

        // then
        expect(pollSpy).toHaveBeenCalledTimes(1);
      });
    });

    describe("usePriority", () => {
      test("should use Priority by default", () => {
        const client = new Client({
          ...customClientOptions,
          autoPoll: false,
        });

        const fetchSpy = jest.spyOn(client.engineService, "fetchAndLock");
        client.subscribe("foo", {}, function () {});
        client.start();

        expect(fetchSpy).toHaveBeenCalled();
        const args = fetchSpy.mock.calls[0][0];
        expect(args.usePriority).toBeTruthy();
      });

      test("should configure priority use", () => {
        const client = new Client({
          ...customClientOptions,
          usePriority: false,
          autoPoll: false,
        });

        const fetchSpy = jest.spyOn(client.engineService, "fetchAndLock");
        client.subscribe("foo", {}, function () {});
        client.start();

        expect(fetchSpy).toHaveBeenCalled();
        const args = fetchSpy.mock.calls[0][0];
        expect(args.usePriority).toBeFalsy();
      });
    });

    describe("when maxParallelExecutions is configured", () => {
      test("should not call fetchAndLock if maximum parallel executions limit exceeded", async () => {
        // given
        const client = new Client({
          ...customClientOptions,
          autoPoll: false,
          maxParallelExecutions: 1,
        });
        const pollSpy = jest.spyOn(client, "poll");

        const engineService = client.engineService;
        engineService.fetchAndLock = jest
          .fn()
          .mockResolvedValue([{ topicName: "foo", variables: {} }]);

        client.subscribe("foo", async () => {
          await new Promise((resolve) =>
            setTimeout(resolve, customClientOptions.interval)
          );
          await new Promise((resolve) =>
            setTimeout(resolve, 0.5 * customClientOptions.interval)
          );
        });

        // when
        client.start();

        // then
        expect(engineService.fetchAndLock).toBeCalled();

        await advanceTimersByTime(customClientOptions.interval);

        expect(pollSpy).toHaveBeenCalledTimes(2);
        expect(engineService.fetchAndLock).toHaveBeenCalledTimes(1);

        await advanceTimersByTime(customClientOptions.interval);

        expect(pollSpy).toHaveBeenCalledTimes(3);
        expect(engineService.fetchAndLock).toHaveBeenCalledTimes(1);
      });
    });

    describe("stop", () => {
      it("should stop polling", async () => {
        // given: we advance time twice then call stop
        const client = new Client(customClientOptions);
        await advanceTimersByTime(2 * customClientOptions.interval);
        client.stop();
        const pollSpy = jest.spyOn(client, "poll");

        // when
        await advanceTimersByTime(2 * customClientOptions.interval);

        // then
        expect(pollSpy).not.toHaveBeenCalled();
      });
    });
  });

  describe("subscribe", () => {
    test("should throw error if api baseUrl wasn't passed as parameter", () => {
      expect(() => {
        new Client();
      }).toThrowError(MISSING_BASE_URL);
    });

    let client, fooWork, customConfig;

    beforeEach(() => {
      client = new Client(customClientOptions);
      fooWork = () => "foobar";
      customConfig = {
        lockDuration: 3000,
        variables: ["fooVariable", "barVariable"],
        processDefinitionId: "processId",
        processDefinitionIdIn: ["processId", "processId2"],
        processDefinitionKey: "processKey",
        processDefinitionKeyIn: ["processKey", "processKey2"],
        processDefinitionVersionTag: "versionTag",
        tenantIdIn: ["tenantId"],
        withoutTenantId: true,
        localVariables: true,
        includeExtensionProperties: true,
      };
    });

    test("should overwrite default with custom client config", () => {
      expect(client.options.workerId).toBe(customClientOptions.workerId);
      expect(client.options.maxTasks).toBe(customClientOptions.maxTasks);
      expect(client.options.interval).toBe(customClientOptions.interval);
      expect(client.options.lockDuration).toBe(
        customClientOptions.lockDuration
      );
    });

    test("should subscribe to topic without custom config ", () => {
      // given
      const footopicSubscription = client.subscribe("foo", fooWork);

      // then
      expect(footopicSubscription.handler).not.toBeUndefined();
    });

    test("should subscribe to topic with custom lockDuration config ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );

      // then
      expect(footopicSubscription.lockDuration).toBe(customConfig.lockDuration);
    });

    test("should subscribe to topic with custom variable subset ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.variables).toBe(customConfig.variables);
    });

    test("should subscribe to topic based on processDefinitionId ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.processDefinitionId).toBe(
        customConfig.processDefinitionId
      );
    });

    test("should subscribe to topic based on processDefinitionIdIn ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.processDefinitionIdIn).toBe(
        customConfig.processDefinitionIdIn
      );
    });

    test("should subscribe to topic based on processDefinitionKey ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.processDefinitionKey).toBe(
        customConfig.processDefinitionKey
      );
    });

    test("should subscribe to topic based on processDefinitionKeyIn ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.processDefinitionKeyIn).toBe(
        customConfig.processDefinitionKeyIn
      );
    });

    test("should subscribe to topic based on processDefinitionVersionTag ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.processDefinitionVersionTag).toBe(
        customConfig.processDefinitionVersionTag
      );
    });

    test("should subscribe to topic based on tenantIdIn ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.tenantIdIn).toBe(customConfig.tenantIdIn);
    });

    test("should subscribe to topic based on withoutTenantId ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.withoutTenantId).toBe(
        customConfig.withoutTenantId
      );
    });

    test("should subscribe to topic based on localVariables ", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.localVariables).toBe(
        customConfig.localVariables
      );
    });

    test("should subscribe to topic with includeExtensionProperties = true", () => {
      // given
      const footopicSubscription = client.subscribe(
        "foo",
        customConfig,
        fooWork
      );
      // then
      expect(footopicSubscription.includeExtensionProperties).toBe(
        customConfig.includeExtensionProperties
      );
    });

    test("should call the API with the custom configs", () => {
      // given
      const fetchAnLockSpy = jest.spyOn(client.engineService, "fetchAndLock");
      client.subscribe("foo", customConfig, fooWork);
      client.poll();

      // then
      expect(fetchAnLockSpy).toBeCalled();
      const args = fetchAnLockSpy.mock.calls[0];
      expect(args).toMatchSnapshot();
    });

    test("should throw error if try to subscribe twice", () => {
      // given
      client.subscribe("foo", fooWork);

      // then
      expect(() => {
        client.subscribe("foo", fooWork);
      }).toThrowError(ALREADY_REGISTERED);
    });

    test("should throw error if handler is not passed", () => {
      expect(() => {
        client.subscribe("foo2");
      }).toThrowError(MISSING_HANDLER);
    });

    test("should allow to unsubscribe from a topic", () => {
      // given
      const footopicSubscription = client.subscribe("foo", fooWork);

      // when
      footopicSubscription.unsubscribe();

      // then
      expect(client.topicSubscriptions["foo"]).toBeUndefined();
    });
  });

  describe("interceptors", () => {
    it("should not add interceptors if they are not provided as a function or array of functions", () => {
      // given
      const options = { ...customClientOptions, interceptors: [] };

      // then
      expect(() => new Client(options)).toThrowError(WRONG_INTERCEPTOR);
    });

    it("should add interceptors if they are provided as a function", () => {
      // given
      const foo = () => {};
      const expectedInterceptors = [foo];
      const options = { ...customClientOptions, interceptors: foo };
      const client = new Client(options);

      // then
      expect(client.engineService.interceptors).toEqual(expectedInterceptors);
    });

    it("should add interceptors if they are provided as an array of functions", () => {
      // given
      const foo = () => {};
      const expectedInterceptors = [foo];
      const options = { ...customClientOptions, interceptors: [foo] };
      const client = new Client(options);

      // then
      expect(client.engineService.interceptors).toEqual(expectedInterceptors);
    });
  });

  describe("middlewares", () => {
    it("should not add middlewares if they are not provided as a function or array of functions", () => {
      // given
      const options = { ...customClientOptions, use: [] };

      // then
      expect(() => new Client(options)).toThrowError(WRONG_MIDDLEWARES);
    });

    it("should accept middleware function", () => {
      // givena
      const middleware = jest.fn();
      const options = { ...customClientOptions, use: middleware };
      const client = new Client(options);

      // then
      expect(middleware).toBeCalledWith(client);
    });

    it("should accept middlewares array", () => {
      // given
      const middlewares = [jest.fn(), jest.fn()];
      const options = { ...customClientOptions, use: middlewares };
      const client = new Client(options);

      // then
      middlewares.forEach((middleware) =>
        expect(middleware).toBeCalledWith(client)
      );
    });
  });

  it("executeTask should call handler with task and taskService", () => {
    // given
    Variables.mockClear();

    let client = new Client({ ...customClientOptions, autoPoll: false });
    const handler = jest.fn();
    client.subscribe("foo", handler);

    const expectedTask = { topicName: "foo", variables: {} };

    // when
    client.executeTask(expectedTask);

    // then
    // Variable service is called with specific parameters (c.f. snapshot)
    const VariablesCall = Variables.mock.calls[0];
    expect(VariablesCall).toMatchSnapshot();
    // handler should be called with specific parameters
    expect(handler).toBeCalled();
    const { task, taskService } = handler.mock.calls[0][0];
    expect(taskService).toBeInstanceOf(TaskService);
    expect(task).toBeDefined();
    expect(task.topicName).toBe(expectedTask.topicName);
    expect(task.variables).toBeInstanceOf(Variables);
  });
});
