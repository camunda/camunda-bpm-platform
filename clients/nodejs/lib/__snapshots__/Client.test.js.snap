// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client executeTask should call handler with task and taskService 1`] = `
[
  {},
  {
    "engineService": EngineService {
      "baseUrl": "http://localhost:XXXX/engine-rest",
      "complete": [Function],
      "fetchAndLock": [Function],
      "get": [Function],
      "handleFailure": [Function],
      "interceptors": undefined,
      "lock": [Function],
      "post": [Function],
      "request": [Function],
      "workerId": "foobarId",
    },
    "processInstanceId": undefined,
    "readOnly": true,
  },
]
`;

exports[`Client subscribe should call the API with the custom configs 1`] = `
[
  {
    "maxTasks": 3,
    "sorting": [
      {
        "sortBy": "createTime",
        "sortOrder": "asc",
      },
    ],
    "topics": [
      {
        "includeExtensionProperties": true,
        "localVariables": true,
        "lockDuration": 3000,
        "processDefinitionId": "processId",
        "processDefinitionIdIn": [
          "processId",
          "processId2",
        ],
        "processDefinitionKey": "processKey",
        "processDefinitionKeyIn": [
          "processKey",
          "processKey2",
        ],
        "processDefinitionVersionTag": "versionTag",
        "tenantIdIn": [
          "tenantId",
        ],
        "topicName": "foo",
        "variables": [
          "fooVariable",
          "barVariable",
        ],
        "withoutTenantId": true,
      },
    ],
    "usePriority": true,
  },
]
`;
